### PR TITLE
Quick round of CI improvements

### PR DESCRIPTION
**Pull request by joshuagl**
Fixes: #455 

### Description of the changes being introduced by the pull request:

* Pin GitHub Actions by digest (#455)
* Remove use of deprecated `set-output` command
* Include latest Python 3.11 in the test matrix
* Switch more of our tox environments to run on the latest Python (but not the linter, see https://github.com/secure-systems-lab/securesystemslib/issues/458)
* Drop permissions for each workflow (h/t @jku)

### Please verify and check that the pull request fulfils the following requirements:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -42,7 +42,7 @@ jobs:
 
       - name: Find pip cache dir
         id: pip-cache
-        run: echo "::set-output name=dir::$(pip cache dir)"
+        run: echo "dir=$(pip cache dir)" >> $GITHUB_OUTPUT
```
Comment (jku) [inline .github/workflows/ci.yml:47]: Nit: this step and the next one could be removed: setup-python now supports pip cache using with-arguments:
```
          cache: 'pip'
          cache-dependency-path: 'requirements*.txt'
```

**Review by jku — APPROVED** [review]
nice

File: .github/workflows/ci.yml (hunk 1 above)
Comment (joshuagl) [reply .github/workflows/ci.yml]: Oh, nice. Thanks for pointing this out! I just pushed an extra patch to make this change.

**Review by jku — APPROVED** [review]

**Review by lukpueh — APPROVED** [review]
Thanks!

**Comment by lukpueh** [comment]
> These changes result in the name of the 'check' changing and therefore breaking our branch protection policy.

Confirm that the jobs pass under their new names. Will update settings.

**Pull request merged**